bgpv1: Add the documentation for using serviceAdvertisements (#31331)
cilium-team/cilium: Documentation/network/bgp-control-plane.rst

Advertising Service Virtual IPs
-------------------------------

In Kubernetes, a Service has multiple virtual IP addresses,
such as ``.spec.clusterIP``, ``.spec.clusterIPs``, ``.status.loadBalancer.ingress[*].ip``
and ``.spec.externalIPs``.
The BGP control plane can advertise the virtual IP address of the Service to BGP peers.

> This allows users to directly access th e Service from outside the cluster.

Reviewer: ```suggestion
This allows users to directly access the Service from outside the cluster.
```

To advertise the virtual IPs, specify the ``virtualRouters[*].serviceSelector`` field
and the ``virtualRouters[*].serviceAdvertisements`` field. The ``.serviceAdvertisements``
defaults to the ``LoadBalancerIP`` service. You can also specify the ``.serviceAdvertisements``
field to advertise specific service types, with options such as ``LoadBalancerIP``,
``ClusterIP`` and ``ExternalIP``.

The ``.serviceSelector`` field is a label selector that selects Services matching
the specified ``.matchLabels`` or ``.matchExpressions``.

BGP Control Plane can advertise the virtual IPs of the Services to the BGP peers.
This allows the BGP peers to reach the Service directly from

> outside the cluster. When your upstream router supports Equal Cost Multi Path

Reviewer: We have a similar phrase above. Maybe we should consolidate?

```
This allows the BGP peers to reach the Service directly from
outside the cluster.
```

(ECMP), you can use this feature to load balance traffic to the Service across
multiple nodes by advertising the same ingress IPs from multiple nodes.

.. code-block:: yaml

   apiVersion: "cilium.io/v2alpha1"
   kind: CiliumBGPPeeringPolicy
   metadata:
     name: rack0
   spec:
     nodeSelector:
       matchLabels:
         rack: rack0
     virtualRouters:
     - localASN: 64512
       serviceSelector: # <-- select Services to advertise
         matchLabels:
           app: foo
       serviceAdvertisements: # <-- specify the service types to advertise
       - LoadBalancerIP # <-- default
       - ClusterIP      # <-- options
       - ExternalIP     # <-- options
       neighbors:
       - peerAddress: '10.0.0.1/32'
         peerASN: 64512

Advertising ExternalIP Services
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> To access Services outside the cluster, specific BPF parameters need to be enabled,
> see the :ref:`External Access To ClusterIP Services <external_access_to_clusterip_services>` section.

Reviewer: ```suggestion
If you wish to use this together with KubeProxyReplacement <reference here>, please make sure the ExternalIP support is enabled.
```

Sorry, actually the `external access to clusterip services` were not required for ExternalIP.

If you only wish to advertise the ``.spec.externalIPs`` of Service,
you can specify the ``virtualRouters[*].serviceAdvertisements`` field as ``ExternalIP``.

.. code-block:: yaml

   apiVersion: "cilium.io/v2alpha1"
   kind: CiliumBGPPeeringPolicy
   metadata:
     name: rack0
   spec:
     nodeSelector:
       matchLabels:
         rack: rack0
     virtualRouters:
     - localASN: 64512
       serviceSelector: # <-- select Services to advertise
         matchLabels:
           app: foo
       serviceAdvertisements: # <-- specify the service types to advertise
       - ExternalIP
       neighbors:
       - peerAddress: '10.0.0.1/32'
         peerASN: 64512

Advertising ClusterIP Services
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> To access Services outside the cluster, specific BPF parameters need to be enabled,
> see the :ref:`External Access To ClusterIP Services <external_access_to_clusterip_services>` section.

Reviewer: ```suggestion
If you wish to use this together with KubeProxyReplacement <reference here>, specific BPF parameters need to be enabled. See :ref:`External Access To ClusterIP Services <external_access_to_clusterip_services>` section for how to enable it.
```
